staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch.go

```diff
@@ -271,89 +329,104 @@ func handleSliceDiff(key string, originalValue, modifiedValue []interface{}, pat
 			patch[parallelDeletionListKey] = deletionList
 		}
 	default:
-		replacePatchFieldIfNotEqual(key, originalValue, modifiedValue, patch, diffOptions)
+		changed = replacePatchFieldIfNotEqualOrReplaceKeys(key, originalValue, modifiedValue, patch, diffOptions)
 	}
-	return nil
+	return changed, nil
 }
 
-// replacePatchFieldIfNotEqual updates the patch if original and modified are not deep equal
+// replacePatchFieldIfNotEqualOrReplaceKeys updates the patch if original and modified are not deep equal
 // if diffOptions.IgnoreChangesAndAdditions is false.
 // original is the old value, maybe either the live cluster object or the last applied configuration
 // modified is the new value, is always the users new config
-func replacePatchFieldIfNotEqual(key string, original, modified interface{},
-	patch map[string]interface{}, diffOptions DiffOptions) {
+// It returns a boolean var indicating if there are any changes.
+func replacePatchFieldIfNotEqualOrReplaceKeys(key string, original, modified interface{},
+	patch map[string]interface{}, diffOptions DiffOptions) bool {
+	if diffOptions.ReplaceKeys {
+		patch[key] = modified
+	}
 	if diffOptions.IgnoreChangesAndAdditions {
 		// Ignoring changes - do nothing
-		return
+		return false
 	}
 	if reflect.DeepEqual(original, modified) {
 		// Contents are identical - do nothing
-		return
+		return false
 	}
 	// Create a patch to replace the old value with the new one
 	patch[key] = modified
+	return true
 }
 
 // updatePatchIfMissing iterates over `original` when ignoreDeletions is false.
 // Clear the field whose key is not present in `modified`.
 // original is the old value, maybe either the live cluster object or the last applied configuration
 // modified is the new value, is always the users new config
-func updatePatchIfMissing(original, modified, patch map[string]interface{}, diffOptions DiffOptions) {
-	if diffOptions.IgnoreDeletions {
-		// Ignoring deletion - do nothing
-		return
-	}
-	// Add nils for deleted values
+// It returns a boolean var indicating if there are any changes.
+func updatePatchIfMissing(original, modified, patch map[string]interface{}, diffOptions DiffOptions) bool {
```

Comment:
>IIUC, basically when IgnoreDeletions==true we do nothing, and when IgnoreDeletions==false we add nils in the for loop, and ReplaceKeys has no nothing to do with it?

Yes.

But IgnoreDeletions(ID) and ReplaceKeys(RK) determine returning true(T) or false(F) jointly.
The truth table is below, where
- nil means add nils in the map
- IC means If there are Changes, it can be true or false.

| | ID=T | ID=F |
|---|---|---|
| RK=T | IC | nil+IC |
| RK=F | F | nil+IC |